### PR TITLE
ISLANDORA-1837: Add configurable Solr field to link Citation and Thes…

### PR DESCRIPTION
**Pull request by glewisorg**
…is objects to Person entities

* Add variable to configuration form with default set to current value (mods_name_personal_displayForm_mt)

* Use variable in queries for RSS feed, citations tab, and theses tab

**JIRA Ticket**: https://jira.duraspace.org/browse/ISLANDORA-1837

* Other Relevant Links (Google Groups discussion, related pull requests, Release pull requests, etc.)

A related pull request will be made to the islandora_scholar module to add name/nameIdentifier to the default Citation and Thesis forms.

# What does this Pull Request do?

The JIRA ticket notes that the linking of Person entities to Citation and Thesis objects is not well documented and accomplished in code with hard-coded queries of the Solr field mods_name_personal_displayForm_mt.  This pull request makes the Solr field configurable and allows the user to specify which MODS element to use for linking.

# What's new?

This pull request adds a variable to configure the Solr field to use for linking of Person entities to Citation and Thesis objects, with its default value set to mods_name_personal_displayForm_mt.  This pull request replaces the hard-coded value with the variable in queries for RSS feed, citation tab, and theses tab.  

The default behavior should be unchanged.  Users that choose to use a different Solr field should verify that the field exists and that it contains the correct value.

# How should this be tested?

Create a Person object with an identifier.  Create a Citation object and a Thesis object, with that person's identifer defined in the MODS element you want to use for linking the person to the object.  After ingest, confirm the name of the Solr field containing the identifer.  In the Entity module configuration form, under "Solr field for linking content to person entities" replace mods_name_personal_displayForm_mt with the name of the Solr field that contains the identifier.  View the RSS feed, Citations tab, and Theses tab to verify the correct objects are listed.

# Interested parties
@Islandora/7-x-1-x-committers

### CODE REVIEW

**Comment by mjordan** [comment]
@glewisorg Could you please complete all the sections of the PR template?

**Comment by glewisorg** [comment]
Sorry, first time using the template.  Let me know if it's not clear enough.

**Comment by mjordan** [comment]
Looks good. You can replace the boilerplate with content specific to the PR (that is, you can remove "A brief description of what the intended result of the PR will be and/or what problem it solves.", "A description of what steps someone could take to:", etc.). Also, if you don't have a specific person in mind under "Interested parties" such as @mjordan,  use @Islandora/7-x-1-x-committers.

Thanks!

**Comment by bondjimbond** [comment]
Thanks for working on this, @glewisorg! I'll test it out tomorrow.

**Comment by bondjimbond** [comment]
First round of testing:
- Changed the Solr field to mods_name_personal_nameIdentifier_ms
- Edited a Thesis and a Citation object to add a <nameIdentifier> to the <name> element, matching the <identifier type="u1"> in the Person entity
- Made sure that the appropriate Solr field is indeed populated for the objects in question
- Cleared caches just to be sure

Result: No Citation or Thesis appears on the Person's scholar profile page.

Did I miss anything in the config/testing steps?

**Comment by bondjimbond** [comment]
Also, a suggestion: Since so many fields on the admin form require precise Solr field names, it would be nice to add autocomplete to them. Taking a snippet from the Solr settings config form seems to work:
`'#autocomplete_path' => 'islandora_solr/autocomplete_luke',`

**Comment by glewisorg** [comment]
Try setting the Solr field to mods_name_personal_nameIdentifier_mt  (mt instead of ms).  Good idea to add autocomplete to all Solr fields on the form.

**Comment by bondjimbond** [comment]
Cool, that seems to be working. Question is, why won't the _ms option work? That should probably be fixed somehow, and at the very least explicitly documented in the config form.

Another issue:

When you click on the Citations or Theses tabs, then, you get the object's Title and Authors. Expected behaviour (how it displays in the 7.x branch) is that the Authors column displays the authors' full names. In this branch, the Authors column just prints the identifier. 

Does anyone know offhand where exactly the code is for printing those results? I imagine it shouldn't be too difficult to get the correct author label via whichever Solr field we're searching with.

**Comment by glewisorg** [comment]
_ms is a string and queries on strings require a complete string match.  If the search string has any spaces or special characters, either each of those characters need to be escaped with a backslash, or quotation marks need to go around the entire search string.  Not sure if that was the problem for you, but it's the general problem I've seen with searching _ms.
The issue with the identifier in the Authors column is my mistake.  I replaced mods_name_personal_displayForm_mt with the new variable in
https://github.com/Islandora/islandora_solution_pack_entities/blob/7.x/includes/utilities.inc#L217-L241
but I should have left those alone.  I will update the pull request.

**Comment by bondjimbond** [comment]
Cool. Regarding the _ms field: in my case, I used a very simple identifier (bweigel), so no special characters. Something else is wrong with the connection to that type of field.

**Comment by glewisorg** [comment]
I can confirm no match with _ms and a simple identifier.  I would guess something to do with how the query is constructed but I would need more time to investigate.
I added a commit to undo my change to the Authors column.  Also added autocomplete for Solr field names to the admin form.

**Comment by bondjimbond** [comment]
Problem is, now that displayForm is not necessary, it won't necessarily be used, and is probably not ideal for that column in the table anyway. Maybe use namePart instead?

**Comment by bondjimbond** [comment]
@glewisorg If I don't have a displayForm value in the MODS record, my Author column is empty. Is it possible to use the namePart instead? Every author should have a namePart, so it's a more reliable approach.

**Comment by glewisorg** [comment]
You could display mods_name_personal_namePart_family_mt.  It's repeatable in the form, so this would display the last name for each author.  It doesn't appear to be a required field in the form, but it seems reasonable to expect that it would have a value.  I don't think I can pair it up with mods_name_personal_namePart_given_mt as that comes back as a separate array.
Another option is to skip Solr fields and use the Coins class from islandora_scholar.  It formats authors as first initial and last name.  Then the Citations tab Authors column would match the Authors row from the Citation object display.

**Comment by bondjimbond** [comment]
dc.contributor works, but it will catch all roles rather than just Author. General question to @Islandora/7-x-1-x-committers - is it possible to call on just the members of the dc.contributor Solr field tagged with (author)?

**Comment by whikloj** [comment]
@glewisorg / @bondjimbond I'm just going to chime in here, it sounds like you are starting to have to guess what end-users are doing. If that is the case I'd make it configurable, because everyone is slightly different. 

@glewisorg I'm not sure why you reversed using the customized field in the returned results. I think that was a valid choice. But if not then perhaps you can just display using the (already) customized _full name_ field (`islandora_entities_disambiguated_solr_field`)?

**Comment by whikloj** [comment]
@bondjimbond as for your `dc.contributor` question, probably but possibly not in a way you'd like. 

But perhaps I don't understand the question. You want to get all the people in the `dc.contributor` Solr field that have been _tagged_ as author. How have they been tagged?

**Comment by bondjimbond** [comment]
@whikloj The mods_to_dc XSLT puts MODS names into dc:contributor, along with their roleTerms in parentheses. So the idea, I suppose, would be to display any dc.contributor fields where the contents include the string "(author)".

**Comment by bondjimbond** [comment]
@glewisorg @whikloj Re the configurable comment: I fully agree that this field should also be configurable. The fewer hard-coded Solr fields we have, the better.

**Comment by glewisorg** [comment]
Do you want me to make the title under the Citations/Theses tabs configurable as well?

**Comment by bondjimbond** [comment]
I think Title can stay as it is. Every object will have a dc.title.

**Comment by bondjimbond** [comment]
Configurable field seems to work. It's hard to find a viable alternative to displayForm for displaying those author names, but then if your repository is using something like nameIdentifier for linking authors to papers, then displayForm can actually be used for its intended purpose (i.e. the way you want the person's name to be displayed).

I think it's good, but I'll leave it open for more @Islandora/7-x-1-x-committers to comment.

**Comment by bondjimbond** [comment]
@rosiel @bryjbrown  This also addresses https://jira.duraspace.org/browse/ISLANDORA-2063

**Review by bondjimbond — COMMENTED** [review]
Looks good to me.

**Review by bondjimbond — APPROVED** [review]

**Comment by bryjbrown** [comment]
@bondjimbond If we could get this merged, it would also fix https://jira.duraspace.org/browse/ISLANDORA-1837. Unfortunately if we want this to appear in the 1.10 release, we'll need to duplicate this PR on the 1.10 branch too. This would violate the sacred Code Freeze tho since this is a new feature and not a bug, but considering how long this PR has sat I'd hope we could make an exception :+1:

**Comment by bondjimbond** [comment]
@bryjbrown Given that it actually does fix a bug that we discussed yesterday (ISLANDORA-2063) -- didn't we say we wanted to fix it for 1.10?

**Comment by bryjbrown** [comment]
@bondjimbond True! Since the 1.10 branch is only ahead by one commit (which was just @DiegoPino doing a version bump), we MAY be able to merge this into 7.x and then merge 7.x to the 1.10 branch? Could be easier that creating a separate PR, but requires more git fu. I'll defer to @DiegoPino @rosiel and @dannylamb on the best choice here.

**Comment by DiegoPino** [comment]
@bryjbrown @bondjimbond we need actually two different pull requests. This one for 7.x, same changes applied to a copy (fork) of 7.x-1.10 as a different pull to get it into the release. Every bug fix during this release times gets the same treatment, not much of a choice 😬 
@rosiel proposed this option with cherry pick (see at the end) https://github.com/Islandora/islandora/wiki/Git-Workflow-Suggestions
there is also a simplyfied version... i do this sometimes... (carefulll!!!!!)
Switch/stay at your current Open branch: glewisorg:7.x-ISLANDORA-1837
Open all the files that are affected by your pull  in an text editor like Textmate, Atom, etc. 
```Shell
git fetch Islandora/7.x-1.10 
// (with "Islandora" being the remote that points to https://github.com/Islandora/islandora_solution_pack_entities/
git checkout 7.x-1.10 
// which will link/track the remote 7.x-1.10 to a local one (creating the local one)
git checkout -b 7.x-1.10-ISLANDORA-1837 
// that will create your actual new pull request branch. Right now just  a copy of the one we have
```
And now the fun part!!
on every open file in textmate, atom, etc. DO A CTRL-Z or CMD-Z (windows/OSX) depending on your OS you could need 1 or 2 (because of the two checkouts) until you see the changes you already commited appear. And save the files 😸  That will bring all the changes into your new branch..
Now you have the chance the make a new commit and a new pull 😁 

I need to make that workflow into a t-shirt jajaja 😆 

UPDATE: this applies to pulls where you are the owner (7.x and release pull). Neglected to see that this pull was from a different user.

**Comment by bondjimbond** [comment]
Attempting @DiegoPino's suggestion with #122

File: includes/entities_rss.inc
Comment (rosiel) [inline includes/entities_rss.inc:74]: Regarding how it only works when searching an `_mt` field, the answer is this: we're searching, for example, `mods_name_personal_naemPart_family_mt:"(dvader)"`. The brackets are a holdover from how the linking used to work (and even then, they weren't actually effective). When searching an _mt field, I think the brackets are dropped as part of the tokenization.  It would be superior to remove the brackets, which would allow us to use an _ms field (if desired) to ensure a 100% match that the fields are identical. The way it is right now and the way it is in this pull request, if your identifiers include spaces then you could get false matches, e.g. "Martin Luther" matches "Martin Luther King".

File: includes/utilities.inc
Comment (rosiel) [inline includes/utilities.inc:249]: Regarding ISLANDORA-2063. This PR does not solve the errors that are appearing on the citation tab. Those errors were introduced when we added the IslandoraSolrQueryProcessor to this module (ISLANDORA-1594) and that ticket has been reopened, and another PR should be on its way shortly! They are because the field that you're looking for (formerly mods_name_personal_displayForm_mt, and/or dc.title) are not in the list of fields you configured solr searches to display, and you configured Solr to "only return those fields" (this is a huge design flaw IMO). 

Also, annoyingly, dc.title (line 248) is never used. This function is only called on line 47 of citation_tab.inc, and there they use the object's label instead of the title that we found here as the text of the link. So the `'title'` line could technically vanish altogether.

File: includes/utilities.inc
Comment (rosiel) [inline includes/utilities.inc:131]: This is outside the range of this PR, but in the same vein of de-hard-coding solr fields, and not a required change for this PR: RELS_EXT_hasModel_uri_mt could (should?) be replaced by the variable islandora_solr_content_model_field, and the pids of the content models ("ir:thesisCModel") replaced by their URIs ("info:fedora/ir:thesisCModel"). This is already used in this module, in line 37 of utilities.inc, though see also the requisite nasty string formatting from lines 58-63.

**Review by rosiel — CHANGES_REQUESTED** [review]
Awesome work! This is going to be a huge improvement, unfortunately the code it touches was messy to begin with.  The only change I'm firmly requesting is the one about brackets in the search string.

Since it was so poorly documented, here's how it works before this PR:
 
In the built-in MADS form there's a (non-mandatory) field called Identifier; it writes to mads/identifier[@type="u1"]. Example identifier: "dvader". 

In the MODS form for Scholar theses/citations there's a field in the name section called "Qualified Name" and it writes to mods/name[@type="personal"]/displayForm. The Entities module hijacks the form and adds an autocomplete to this element that points to 'islandora/entities/autocomplete/disambiguated'. 

That function (which is defined in callbacks.inc) searches what you're typing into the form against the solr field MADS_disambiguated_fullname_mt (or whatever you've configured in the variable islandora_entities_disambiguated_solr_field). The de-facto default transform for MADS, which was written by DGI, creates a special Disambiguated field out of the concatenated GIVEN and FAMILY names plus the U1 identifier in brackets. e.g. "Darth Vader (dvader)". 

So currently the query is searching MADS_disambiguated_fullname_mt:"(dvader)" which because _mt is a tokenized field, while the brackets happen to be there in the target they don't actually matter. But this PR could actually improve the functionality if the brackets were removed from the search string which would allow you to potentially use a field that contained nothing but the identifier.

**Comment by DiegoPino** [comment]
@rosiel @bondjimbond see https://github.com/Islandora/islandora_solution_pack_entities/pull/123#issuecomment-331753150

**Comment by DiegoPino** [comment]
@bondjimbond 
After some talks with Rosie, i see there are a few issues with this pull:

A) it tries to solve 2 JIRA tickets, one is an improvement (ISLANDORA-1837) and the other a Bug fix (ISLANDORA-2023). We can only merge into the Release if it deals with the Bug fix (ISLANDORA-2023): so all the code that applies the improvement needs to be removed. Sorry, rules are rules. 

B) the bug fix is, according to @rosiel comments, not being addressed. I would ask @rosiel that opened a pull request that [deals with ISLANDORA-1587 which is basically the same issue](https://github.com/Islandora/islandora_solution_pack_entities/pull/123) to help you get the fix done so this pull gets into the release.

In conclusion: 
- remove any reference to 1837 (pull message, title and code)
- get the actual fix into the code. With 1. your own code or 2.  looking at https://github.com/Islandora/islandora_solution_pack_entities/pull/123,  or with @rosiel's help. Whatever makes everyone happier.

2023, if fixed right will also close 1594. 
Once that happens, https://github.com/Islandora/islandora_solution_pack_entities/pull/123 can be closed. I would like to make this one the fix. If we all work together that can be done for sure.

@bryjbrown ping. What do you think? @rosiel @bondjimbond any other concerns?

Thanks a lot, go team 😄

**Comment by DiegoPino** [comment]
@glewisorg sorry for the whole confusion here. Somehow this improvement pull request got mixed up with Jira ticket https://jira.duraspace.org/browse/ISLANDORA-2063 and a pull derived from yours attributed to be the fix. I think we all got confused and mixed up in the road.
This pull request is an actually good one and needed improvement

#122 will need to be fixed and changed if it actually wants to fix ISLANDORA-2063 and your code removed from there to avoid mixing up stuff and taking attributions of your authorship

@bondjimbond @bryjbrown @rosiel do you agree with that? @rosiel can you drop the change requests? Let's deal with ISLANDORA-2063 on #122 and lets go from there

thanks, and again, sorry for this whole confusion

**Review by rosiel — APPROVED** [review]

**Comment by rosiel** [comment]
Sorry bout the confusion. This *is* a good improvement, and works as advertised, so any improvements can be done after the fact. Good to merge!

Its equivalent on 7.x-1.10, though, #122, cannot be merged (not as is). We'll deal with the bug ISLANDORA-2063 in #123 / #124 (or, if you prefer, with a significant rewrite of #122). It was rude of me to open a separate PR when there was one already in progress. I'm sorry.

**Comment by bondjimbond** [comment]
Cool cool... I'll hit the Merge button this afternoon after the requisite 24 hours (unless someone beats me to it). @rosiel I closed #122, so how should we go about getting a 7.x-1.10 equivalent if @glewisorg isn't planning on taking it on for the upcoming release?

**Comment by DiegoPino** [comment]
@bondjimbond thanks. There won't be a #120 into the release. Improvements are only allowed to go into the release before code freeze.

**Pull request merged**
